cpp/src/groupby/sort/scan.cpp

```diff
@@ -115,51 +121,91 @@ template <>
+  auto const& group_labels     = helper.group_labels(stream);
+  auto const group_labels_view = column_view(cudf::device_span<const size_type>(group_labels));
+  auto const gather_map =
+    is_presorted()
+      ? cudf::detail::sequence(
+          group_labels.size(), *cudf::make_fixed_width_scalar(size_type{0}, stream), stream)
+      : ((rank_agg._method == rank_method::FIRST
+            ? cudf::detail::stable_sorted_order
+            : cudf::detail::sorted_order)(table_view({group_labels_view, get_grouped_values()}),
+                                          {order::ASCENDING, rank_agg._column_order},
+                                          {null_order::AFTER, rank_agg._null_precedence},
+                                          stream,
+                                          rmm::mr::get_current_device_resource()));
+
+  auto rank_scan = [&]() {
+    switch (rank_agg._method) {
+      case rank_method::FIRST: return detail::first_rank_scan;
+      case rank_method::AVERAGE: return detail::average_rank_scan;
+      case rank_method::DENSE: return detail::dense_rank_scan;
+      case rank_method::MIN: return detail::min_rank_scan;
+      case rank_method::MAX: return detail::max_rank_scan;
+      default: CUDF_FAIL("Unsupported rank method in groupby scan");
+    }
+  }();
+  auto result = rank_scan(get_grouped_values(),
+                          *gather_map,
+                          helper.group_labels(stream),
+                          helper.group_offsets(stream),
+                          stream,
+                          rmm::mr::get_current_device_resource());
+  if (rank_agg._percentage) {
+    auto count = get_grouped_values().nullable() and rank_agg._null_handling == null_policy::EXCLUDE
+                   ? detail::group_count_valid(get_grouped_values(),
+                                               helper.group_labels(stream),
+                                               helper.num_groups(stream),
+                                               stream,
+                                               rmm::mr::get_current_device_resource())
+                   : detail::group_count_all(helper.group_offsets(stream),
+                                             helper.num_groups(stream),
+                                             stream,
+                                             rmm::mr::get_current_device_resource());
+    result     = detail::group_rank_to_percentage(rank_agg._method == rank_method::DENSE,
+                                              *result,
+                                              *count,
+                                              helper.group_labels(stream),
+                                              helper.group_offsets(stream),
+                                              stream,
+                                              mr);
+  }
+  result = std::move(cudf::detail::scatter(
```

Comment:
Likely a silly question: How come this is a `scatter()` call, and not a `gather()`?

Reply:
I am using gather map from sorted order to restore the rank to original order of grouped values.
I might be missing an easier way to using sorted order gather map, get the original order back. is there an easy way?

Reply:
Only side effect which I don't like is that, `scatter` copies the source column first.

Reply:
Hmm why is that? I wouldn't have expected that to be necessary. Does scatter perform an in-place mutation on the resulting copy?

Reply:
https://github.com/rapidsai/cudf/blob/ca952f8355810c46bb132f0b50a67e4109108ccb/cpp/include/cudf/detail/scatter.cuh#L149
*copies* the source column first, then does inplace mutation.